### PR TITLE
feat(catalog-backend): support filtering on property existance

### PR DESCRIPTION
**Pull request by kuangp**
Signed-off-by: Phil Kuang <pkuang@factset.com>

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by kuangp** [comment]
@freben was there a particular reason why this wasn't supported before? From what I understand, everything was already in place in the actual code to support this feature and from my testing, this worked fine after removing the query string restriction

File: plugins/techdocs/src/home/components/TechDocsCustomHome.tsx
Comment (iamEAP) [inline plugins/techdocs/src/home/components/TechDocsCustomHome.tsx:128]: @benjdlambert @reviewers We as TechDocs folk would naturally love this kinda thing.  But I can also see cases where you might want to be able to differentiate between "this key strictly is an empty string" and "this key is set" and will leave it up to you on how you want to handle that in your API.

Comment (benjdlambert) [reply plugins/techdocs/src/home/components/TechDocsCustomHome.tsx:128]: Yeah, I've not really been over this, but i'd say it could be worth exploring the idea of having `filter` also support an array or set of keys that should be defined rather than this, as you're right it's hard to distinguish between the two

Comment (freben) [reply plugins/techdocs/src/home/components/TechDocsCustomHome.tsx:128]: So supporting these things involves several different places in the code and protocols right.

There's the `CatalogApi`, as in the commented line here. If we intend to keep the `filter` parameter as a simple key-value object, I would probably prefer that `x: ''` means equal to the empty string, and e.g. `x: CatalogApi.EXISTS` which is a Symbol means to check for existence. Something like that?

Then there's the REST API. I think that the syntax `?filter=x=` should continue to mean equal to the empty string, while e.g. `?filter=x` could mean that it's defined and has any value. Maybe?

Then there's the [`EntityFilter`](https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/database/types.ts#L129) that this all gets translated to. I would expect that [`matchValueIn`](https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/database/types.ts#L120) would be amended with a `matchValueExists: true` or something. That would also be possible to extend in the future with a `matchValueExists: false` which requires that it's NOT present.

Then there's the database layer, where this all gets [translated](https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/next/NextEntitiesCatalog.ts#L79) to actual SQL.

Comment (kuangp) [reply plugins/techdocs/src/home/components/TechDocsCustomHome.tsx:128]: Amending with an extra `matchValueExists` prop seems like it might lead to some ambiguity. 

If `matchValueIn` is defined, then by definition, it means the associated key must exist but if `matchValueExists = true` and `matchValueIn` specifies an array of values, does that translate to just checking if the key exists or that it must also match one of those values? and then theres the case where `matchValueExists = false` but `matchValueIn` is still defined, which one takes precedence? A sensible logic flow might be to first check if `matchValueExists` is defined and query based off that ignoring `matchValueIn` and if it's undefined, fallback to the current behavior querying based on `matchValueIn`.

Alternatively, it might be simpler to go with: if `matchValueIn` is defined, the key must exist and match any of the values and `matchValueIn = []` just checks for existence regardless of value. If `matchValueIn` is undefined, the key must not exist. 

Thoughts?

Comment (freben) [reply plugins/techdocs/src/home/components/TechDocsCustomHome.tsx:128]: It would just redundantly check both. I.e. there would be a SQL fragment that matches for existence and one that matches for values. I'm not concerned about that though, it's a corner case and not harmful. I'd prefer that, over "abusing" the mechanism for values here too. The reader of the code will then have to be aware at all times that "oh ok if the array is empty then i need to make some special case handling of that" since the SQL will definitely be looking different.

Comment (kuangp) [reply plugins/techdocs/src/home/components/TechDocsCustomHome.tsx:128]: hmm, if I understand correctly, would this be the proper way to implement this in the following code? https://github.com/backstage/backstage/blob/e47f2a3c095227245e5ac0048daa6fbea989ea8b/plugins/catalog-backend/src/next/NextEntitiesCatalog.ts#L79-L104
```diff
-       for (const { key, matchValueIn } of singleFilter.allOf) { 
+       for (const { key, matchValueIn, matchValueExists } of singleFilter.allOf) {
          // NOTE(freben): This used to be a set of OUTER JOIN, which may seem to
          // make a lot of sense. However, it had abysmal performance on sqlite
          // when datasets grew large, so we're using IN instead.
          const matchQuery = db<DbSearchRow>('search')
            .select('entity_id')
            .where(function keyFilter() {
              this.andWhere({ key: key.toLowerCase() });
-            if (matchValueIn) { 
+            if (matchValueExists !== false && matchValueIn) {
                if (matchValueIn.length === 1) {
                  this.andWhere({ value: matchValueIn[0].toLowerCase() });
                } else if (matchValueIn.length > 1) {
                  this.andWhere(
                    'value',
                    'in',
                    matchValueIn.map(v => v.toLowerCase()),
                  );
                }
              }
            });
-         this.andWhere('entity_id', 'in', matchQuery); 
+         this.andWhere('entity_id', matchValueExists === false ? 'not in' : 'in', matchQuery);
        }
```
taking into account that either `matchValueIn` and `matchValueExists` may be undefined and handling the different combinations of how these two props may be defined. 


Also for completeness, the implementation for the alternative solution already has the empty array case handled inherently and would only require changing the following line. One could also argue that handling of the empty array case is already ambiguous in what the intended behavior should be and requires some special interpretation
```ts
          this.andWhere('entity_id', matchValueIn ? 'in' : 'not in', matchQuery);
```

Comment (freben) [reply plugins/techdocs/src/home/components/TechDocsCustomHome.tsx:128]: Hm, not in a great position to review in detail now but yes that looks like it's sensible.

Don't worry too much about the edge case of an empty array. Remember that this is populated from parsing a query, and there's no way of having that parser result in a filter set with thar behavior.

**Comment by kuangp** [comment]
@freben @benjdlambert @iamEAP updated this change to distinctly support filtering on empty strings and key existence according to @freben's suggestion. Also implemented support for key non-existence filtering on the backend side pending how we want to expose querying this from the catalog API in the future

**Comment by benjdlambert** [comment]
Added the `after-vacation` label to this @kuangp as want to get some more eyes on this when the other folks are back from their :beach:

File: packages/catalog-client/src/types/api.ts
Comment (freben) [inline packages/catalog-client/src/types/api.ts:19]:
```suggestion
export const CATALOG_FILTER_EXISTS = Symbol('CATALOG_FILTER_EXISTS');
```

**Review by freben — APPROVED** [review]
Happily approved, looks good! Only one nit that perhaps we could add a name to the symbol so it's even more easily identifiable while debugging etc.

**Comment by kuangp** [comment]
> Added the `after-vacation` label to this @kuangp as want to get some more eyes on this when the other folks are back from their 🏖️

@benjdlambert was wondering if we still need to hold off on this PR since @freben has given his approval or is there a chance this could go out with this weeks release? We are very keen to have the performance improvement for loading techdocs

**Comment by benjdlambert** [comment]
@kuangp ill take a look at this again in the morning and see if I can get it in the release tomorrow

**Review by benjdlambert — APPROVED** [review]

**Comment by benjdlambert** [comment]
Just need a review from @backstage/techdocs-core

**Review by iamEAP — APPROVED** [review]
Sweet!  I think there's opportunity to apply this in the `DefaultTechDocsCollator` too, but we can handle that in a follow-up.  LGTM.

**Pull request merged**
